File: src/main/java/org/trp/shincolle/entity/base/path/ShipLegacyPath.java
package org.trp.shincolle.entity.base.path;

import net.minecraft.world.entity.Entity;
import net.minecraft.world.phys.Vec3;

import javax.annotation.Nullable;

final class ShipLegacyPath {

    private final ShipLegacyPathPoint[] points;
    private int currentPathIndex;

    ShipLegacyPath(ShipLegacyPathPoint[] points) {
        this.points = points;
    }

    int getCurrentPathLength() {
        return this.points.length;
    }

    int getCurrentPathIndex() {
        return this.currentPathIndex;
    }

    void setCurrentPathIndex(int currentPathIndex) {
        this.currentPathIndex = Math.max(0, Math.min(currentPathIndex, this.points.length));
    }

    void incrementPathIndex() {
        setCurrentPathIndex(this.currentPathIndex + 1);
    }

    boolean isFinished() {
        return this.currentPathIndex >= this.points.length;
    }

    ShipLegacyPathPoint getFinalPathPoint() {
        return this.points.length > 0 ? this.points[this.points.length - 1] : null;
    }

    @Nullable
    ShipLegacyPathPoint getPathPointFromIndex(int index) {
        if (index < 0 || index >= this.points.length) {
            return null;
        }

        return this.points[index];
    }

    @Nullable
    Vec3 getCurrentPos() {
        if (isFinished()) {
            return null;
        }

        ShipLegacyPathPoint point = this.points[this.currentPathIndex];
        return new Vec3(point.getX(), point.getY(), point.getZ());
    }

    @Nullable
    Vec3 getVectorFromIndex(Entity entity, int index) {
        if (index < 0 || index >= this.points.length) {
            return null;
        }

        ShipLegacyPathPoint point = this.points[index];
        double x = point.getX() + (double) ((int) (entity.getBbWidth() + 1.0F)) * 0.5D;
        double y = point.getY();
        double z = point.getZ() + (double) ((int) (entity.getBbWidth() + 1.0F)) * 0.5D;
        return new Vec3(x, y, z);
    }

    @Nullable
    Vec3 getPosition(Entity entity) {
        return getVectorFromIndex(entity, this.currentPathIndex);
    }
}

File: src/main/java/org/trp/shincolle/entity/base/path/ShipLegacyPath.kt
package org.trp.shincolle.entity.base.path

import net.minecraft.world.entity.Entity
import net.minecraft.world.phys.Vec3
import kotlin.math.max
import kotlin.math.min

internal class ShipLegacyPath(private val points: Array<ShipLegacyPathPoint>) {
    private var pathIndex = 0

    val currentPathLength: Int
        get() = this.points.size

    val currentPathIndex: Int
        get() = this.pathIndex

    fun setCurrentPathIndex(currentPathIndex: Int) {
        this.pathIndex = max(0, min(currentPathIndex, this.points.size))
    }

    fun incrementPathIndex() {
        setCurrentPathIndex(this.pathIndex + 1)
    }

    val isFinished: Boolean
        get() = this.pathIndex >= this.points.size

    val finalPathPoint: ShipLegacyPathPoint?
        get() = if (this.points.size > 0) this.points[this.points.size - 1] else null

    fun getPathPointFromIndex(index: Int): ShipLegacyPathPoint? {
        if (index < 0 || index >= this.points.size) {
            return null
        }

        return this.points[index]
    }

    val currentPos: Vec3?
        get() {
            if (this.isFinished) {
                return null
            }

            val point = this.points[this.pathIndex]
            return Vec3(point.x.toDouble(), point.y.toDouble(), point.z.toDouble())
        }

    fun getVectorFromIndex(entity: Entity, index: Int): Vec3? {
        if (index < 0 || index >= this.points.size) {
            return null
        }

        val point = this.points[index]
        val x = point.x + ((entity.bbWidth + 1.0f).toInt()).toDouble() * 0.5
        val y = point.y.toDouble()
        val z = point.z + ((entity.bbWidth + 1.0f).toInt()).toDouble() * 0.5
        return Vec3(x, y, z)
    }

    fun getPosition(entity: Entity): Vec3? {
        return getVectorFromIndex(entity, this.pathIndex)
    }
}
